### PR TITLE
Docker Integration

### PR DESCRIPTION
**Pull request by programster**
This pull request allows users to deploy Lychee through docker using the codebase itself, rather than some third party wrapper such as [kdelfour/lychee-docker](https://hub.docker.com/r/kdelfour/lychee-docker/).

[This youtube video demonstrates the code in action](https://www.youtube.com/watch?v=iADJDL9xda4&feature=youtu.be).

## Resons for Development

It looks like the Lychee used to support building through docker, but then using kdelfour/lychee-docker [became the official way](https://github.com/electerious/Lychee/issues/575), but the problem with this is that the third party wrapper may get out of sync with the codebase. For example, if a new branch needs a new package, then its easier for that branch to just edit the Dockerfile here for when it is ready to be merged into master. 

Also, it looks like the kdelfour image is not being kept up-to-date. [The last time an image was pushed was 8 months ago](https://hub.docker.com/r/kdelfour/lychee-docker/builds/). 

Another factor is that the kdelfour image uses php 5.5 instead of this integration which uses 5.6. Php 5.5 isn't supported anymore, and 5.6 only has security updates. It would be good to upgrade to 7.0 or 7.1 which this should make somewhat easier (ties into my first point).

Lastly, the kdelfour image works based on using an internal MySQL server. [As somebody else has pointed out](https://github.com/kdelfour/lychee-docker/issues/10), it's much better practice to keep the database outside the application container, either by using another container, or by just running your own MySQL server. In this case you can do either, and just pass the details in using environment variables.

### CODE REVIEW

**Comment by Wonderfall** [comment]
If I may suggest something, it'd a great idea to move to Alpine Linux as a base, instead of Debian Jessie. More and more official images are migrating to Alpine since it's lightweight and more suitable for containers (in the vast majority of cases).

I'm the author of [wonderfall/lychee](https://github.com/Wonderfall/dockerfiles/tree/master/lychee), if my work can help (it's not a lot a work, I prefer to keep it simple, and it's working flawlessly for my needs).

**Comment by programster** [comment]
@Wonderfall I have seen a lot of people move over to alpine linux, and to be honest, I'm not too fussed about what distribution is used for the container, _as long as we have the docker build inside the project_ rather than outside for the reasons stated in the pull request:

> ...the third party wrapper may get out of sync with the codebase. For example, if a new branch needs a new package, then its easier for that branch to just edit the Dockerfile here for when it is ready to be merged into master.

By having docker integration in a project, there is one known environment to build and test against that is supposed to work. If people decide to deploy in some other manner with different versions of PHP etc, then it is their own fault if something breaks either immediately or in the future after an update.

The main reasons for sticking with a base image like Debian or Ubuntu rather than switching to Alpine, is that its probably easier for people who wish to deploy manually, or set up their own dev environments. I believe a lot of developers are fairly familiar/used to deploying on debian/ubuntu in their day-to-day life, compared to Alpine which I believe is aimed squarely at things like docker containers. They would not have a hard time reproducing their dev environment by reading through the Dockerfile. One can mostly just replace RUN with sudo. This is also good for developers to set up their dev environments for a rapid iteration cycle, rather than having to build their container every time they make a change to the code. Admittedly, they could [dev inside the container using SSHFS](http://programster.blogspot.co.uk/2014/08/dev-inside-docker-container-with-sshfs.html), but even I didn't stick to that. 

The size of the base image doesn't really affect the build process much, but does affect the deployments if users are just pulling from the dockerhub instead of building their own container. However, the difference in the size between the two base images is 51.36MB, and 29.92MB. Yes its nearly twice the size, but only 21.44 MB size difference. 

In conclusion, I believe go ahead with alpine if a lighter weight final docker image is most important, or stick with debian/ubuntu if you want to make it easier for other developers to provide contributions to the project and actually test their changes against the build environment.

**Comment by Wonderfall** [comment]
As I'm not a developer at all, I was only interested in final deployment (and at this perspective I only see advantages for Alpine Linux). Now if you say it's better for dev, I obviously can't say the contrary.

**Comment by programster** [comment]
@Wonderfall sorry, I didn't mean to say that Debian is _better_ than alpine Linux for development. I was just trying to make the case that it might be easier for other people to contribute if the environment is debian/ubuntu based as I _think_ more people are generally more familiar with those distros. I could be wrong and perhaps Alpine linux is way more popular than I realize. A good example of what I'm saying is that I can easily take this pull request further and implement ssl handling, but I'd have a bugger of a time figuring that out with Alpine, but maybe other's can take care of that.

I really don't want to come across as x is better than y. Alpine is clearly a lighter image that is gaining popularity in the docker sphere.

**Comment by SirEdvin** [comment]
Alpine is terrible, when you use something, that unsupported by his community, like RabbitMQ.

**Comment by programster** [comment]
Can we just merge this and if the community decides Debian is not suitable change it as and when? This pull request is better than our current situation. If there is an actual issue with the pull request, then let me know and I will try to address it.

**Comment by donutmonger** [comment]
Why isn't this merged yet? I would love a docker image for lychee.

**Comment by rgarrigue** [comment]
Meanwhile, this one works : https://hub.docker.com/r/linuxserver/lychee/

**Comment by Wonderfall** [comment]
There are multiple working images : https://hub.docker.com/search/?isAutomated=0&isOfficial=0&page=1&pullCount=0&q=lychee&starCount=0

**Comment by CodingFrieda** [comment]
Any updates on this?

**Comment by electerious** [comment]
This PR will be closed because I'm not actively working on Lychee anymore. The repository of Lychee will stay available, but the main development continues under the [Lychee Organisation](https://github.com/LycheeOrg). A fork maintained by the community.
